File: src/lsp.zig
const std = @import("std");

pub const JsonRpcRequest = struct {
    jsonrpc: []const u8 = "2.0",
    id: ?i64,
    method: []const u8,
    params: ?std.json.Value,
};

pub const JsonRpcNotification = struct {
    jsonrpc: []const u8 = "2.0",
    method: []const u8,
    params: ?std.json.Value,
};

pub const JsonRpcResponse = struct {
    id: ?i64,
    result: ?std.json.Value,
    @"error": ?ResponseError,
};

pub const ResponseError = struct {
    code: i64,
    message: []const u8,
};

pub const ServerCapabilities = struct {
    hover_provider: bool = false,
    signature_help_provider: bool = false,
    completion_provider: bool = false,
    definition_provider: bool = false,
    references_provider: bool = false,
    document_symbol_provider: bool = false,
    document_formatting_provider: bool = false,
    code_action_provider: bool = false,
    rename_provider: bool = false,
    inlay_hint_provider: bool = false,
    workspace_symbol_provider: bool = false,
};

/// Parse server capabilities from the initialize response
pub fn parseServerCapabilities(value: std.json.Value) ServerCapabilities {
    var caps = ServerCapabilities{};

    const obj = switch (value) {
        .object => |o| o,
        else => return caps,
    };

    const capabilities = obj.get("capabilities") orelse return caps;
    const caps_obj = switch (capabilities) {
        .object => |o| o,
        else => return caps,
    };

    caps.hover_provider = isTruthy(caps_obj.get("hoverProvider"));
    caps.signature_help_provider = caps_obj.get("signatureHelpProvider") != null;
    caps.completion_provider = caps_obj.get("completionProvider") != null;
    caps.definition_provider = isTruthy(caps_obj.get("definitionProvider"));
    caps.references_provider = isTruthy(caps_obj.get("referencesProvider"));
    caps.document_symbol_provider = isTruthy(caps_obj.get("documentSymbolProvider"));
    caps.document_formatting_provider = isTruthy(caps_obj.get("documentFormattingProvider"));
    caps.code_action_provider = isTruthy(caps_obj.get("codeActionProvider"));
    caps.rename_provider = isTruthy(caps_obj.get("renameProvider"));
    caps.inlay_hint_provider = isTruthy(caps_obj.get("inlayHintProvider"));

    if (caps_obj.get("workspace")) |ws| {
        if (ws == .object) {
            if (ws.object.get("symbol")) |sym| {
                caps.workspace_symbol_provider = sym != .null;
            }
        }
    }

    return caps;
}

fn isTruthy(v: ?std.json.Value) bool {
    const val = v orelse return false;
    return switch (val) {
        .bool => |b| b,
        .object => true,
        .null => false,
        else => false,
    };
}

/// LSP Client manages the subprocess and read/write buffers
pub const Client = struct {
    allocator: std.mem.Allocator,
    child: std.process.Child,
    next_id: i64,
    read_buf: std.ArrayList(u8),

    pub fn init(allocator: std.mem.Allocator, server_path: []const u8, server_args: []const []const u8) !Client {
        var argv: std.ArrayList([]const u8) = .empty;
        defer argv.deinit(allocator);
        try argv.append(allocator, server_path);
        for (server_args) |arg| try argv.append(allocator, arg);

        var child = std.process.Child.init(argv.items, allocator);
        child.stdin_behavior = .Pipe;
        child.stdout_behavior = .Pipe;
        child.stderr_behavior = .Ignore;

        try child.spawn();

        const empty_arraylist: std.ArrayList(u8) = .empty;
        return Client{
            .allocator = allocator,
            .child = child,
            .next_id = 1,
            .read_buf = empty_arraylist,
        };
    }

    pub fn deinit(self: *Client) void {
        // Send shutdown/exit before killing
        if (self.child.stdin) |stdin| {
            _ = stdin;
        }
        _ = self.child.kill() catch {};
        self.read_buf.deinit(std.heap.page_allocator);
    }

    pub fn sendRequest(self: *Client, method: []const u8, params: ?std.json.Value) !i64 {
        const id = self.next_id;
        self.next_id += 1;

        var buf: std.ArrayList(u8) = .empty;
        defer buf.deinit(std.heap.page_allocator);

        const writer = buf.writer(std.heap.page_allocator);
        try writer.writeAll("{\"jsonrpc\":\"2.0\",\"id\":");
        try writer.writeAll(try std.fmt.allocPrint(std.heap.page_allocator, "{d}", .{id}));
        try writer.writeAll(",\"method\":\"");
        try writer.writeAll(method);
        try writer.writeAll("\"");

        if (params) |p| {
            try writer.writeAll(",\"params\":");
            const fmter = std.json.fmt(p, .{});
            try writer.writeAll(try std.fmt.allocPrint(std.heap.page_allocator, "{f}", .{fmter}));
        } else {
            try writer.writeAll(",\"params\":null");
        }
        try writer.writeAll("}");

        try self.sendRaw(buf.items);
        return id;
    }

    pub fn sendNotification(self: *Client, method: []const u8, params: ?std.json.Value) !void {
        var buf: std.ArrayList(u8) = .empty;
        defer buf.deinit(std.heap.page_allocator);

        var writer = buf.writer(std.heap.page_allocator);

        try writer.print("{{\"jsonrpc\":\"2.0\",\"method\":\"{s}\"", .{method});

        if (params) |p| {
            try writer.print(",\"params\":{f}", .{std.json.fmt(p, .{})});
        }

        try writer.writeAll("}");

        try self.sendRaw(buf.items);
    }

    fn sendRaw(self: *Client, content: []const u8) !void {
        const stdin = self.child.stdin orelse return error.NoStdin;
        var header_buf: [64]u8 = undefined;
        const header = try std.fmt.bufPrint(&header_buf, "Content-Length: {d}\r\n\r\n", .{content.len});
        try stdin.writeAll(header);
        try stdin.writeAll(content);
    }

    /// Read one JSON-RPC message, parse it. Caller owns the returned value's arena.
    pub fn readMessage(self: *Client, timeout_ms: u64) !?ParsedResponse {
        const stdout = self.child.stdout orelse return error.NoStdout;

        const deadline = std.time.milliTimestamp() + @as(i64, @intCast(timeout_ms));

        // Read until we have a complete message
        while (true) {
            if (std.time.milliTimestamp() > deadline) return null;

            // Try to parse existing buffer
            if (try self.tryParseMessage()) |msg| return msg;

            const remaining_ms = deadline - std.time.milliTimestamp();
            if (remaining_ms <= 0) return null;

            var tmp: [4096]u8 = undefined;
            const n = stdout.read(&tmp) catch return null;
            if (n == 0) return null;
            try self.read_buf.appendSlice(std.heap.page_allocator, tmp[0..n]);
        }
    }

    fn tryParseMessage(self: *Client) !?ParsedResponse {
        const data = self.read_buf.items;

        // Find the header/body separator
        const sep = std.mem.indexOf(u8, data, "\r\n\r\n") orelse return null;
        const header = data[0..sep];

        // Parse Content-Length
        const cl_prefix = "Content-Length: ";
        const cl_start = std.mem.indexOf(u8, header, cl_prefix) orelse return null;
        const cl_val_start = cl_start + cl_prefix.len;
        const cl_end = std.mem.indexOfScalarPos(u8, header, cl_val_start, '\r') orelse header.len;
        const content_length = try std.fmt.parseInt(usize, std.mem.trim(u8, header[cl_val_start..cl_end], " \r\n"), 10);

        const body_start = sep + 4;
        if (data.len < body_start + content_length) return null;

        const body = data[body_start .. body_start + content_length];

        // Parse JSON
        var arena = std.heap.ArenaAllocator.init(self.allocator);
        errdefer arena.deinit();
        const parsed = try std.json.parseFromSliceLeaky(std.json.Value, arena.allocator(), body, .{});

        // Consume the bytes
        const consumed = body_start + content_length;
        try self.read_buf.replaceRange(std.heap.page_allocator, 0, consumed, &.{});

        return ParsedResponse{ .arena = arena, .value = parsed };
    }

    /// Read the response to a specific request ID, skipping notifications
    pub fn readResponse(self: *Client, id: i64, timeout_ms: u64) !?ParsedResponse {
        const deadline = std.time.milliTimestamp() + @as(i64, @intCast(timeout_ms));

        while (std.time.milliTimestamp() < deadline) {
            const remaining = @as(u64, @intCast(@max(0, deadline - std.time.milliTimestamp())));
            const msg = try self.readMessage(@min(remaining, 200)) orelse return null;

            // Check if this is the response we want
            const obj = switch (msg.value) {
                .object => |o| o,
                else => {
                    msg.arena.deinit();
                    continue;
                },
            };

            // Skip notifications (no id field or id is null)
            const msg_id = obj.get("id") orelse {
                msg.arena.deinit();
                continue;
            };

            const msg_id_int = switch (msg_id) {
                .integer => |i| i,
                else => {
                    msg.arena.deinit();
                    continue;
                },
            };

            if (msg_id_int == id) return msg;
            msg.arena.deinit();
        }
        return null;
    }
};

pub const ParsedResponse = struct {
    arena: std.heap.ArenaAllocator,
    value: std.json.Value,

    pub fn deinit(self: *ParsedResponse) void {
        self.arena.deinit();
    }

    pub fn hasError(self: *const ParsedResponse) bool {
        const obj = switch (self.value) {
            .object => |o| o,
            else => return false,
        };
        return obj.get("error") != null;
    }

    pub fn getResult(self: *const ParsedResponse) ?std.json.Value {
        const obj = switch (self.value) {
            .object => |o| o,
            else => return null,
        };
        return obj.get("result");
    }
};

File: src/lsp.rs
use memchr::memmem;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::io::{Read, Write};
use std::process::{Child, Command, Stdio};
use std::time::{Duration, Instant};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcRequest {
    pub jsonrpc: String,
    pub id: i64,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JsonRpcNotification {
    pub jsonrpc: String,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Clone, Default)]
pub struct ServerCapabilities {
    pub hover_provider: bool,
    pub signature_help_provider: bool,
    pub completion_provider: bool,
    pub definition_provider: bool,
    pub type_definition_provider: bool,
    pub implementation_provider: bool,
    pub references_provider: bool,
    pub document_symbol_provider: bool,
    pub workspace_symbol_provider: bool,
    pub document_formatting_provider: bool,
    pub code_action_provider: bool,
    pub rename_provider: bool,
    pub prepare_rename_provider: bool,
    pub inlay_hint_provider: bool,
    pub code_lens_provider: bool,
    pub semantic_tokens_provider: bool,
    pub folding_range_provider: bool,
    pub linked_editing_range_provider: bool,
    pub selection_range_provider: bool,
    pub document_highlight_provider: bool,
    pub publish_diagnostics_provider: bool,
    pub execute_command_provider: bool,
    pub did_change_configuration_provider: bool,
    pub did_change_workspace_folders_provider: bool,
    pub text_document_sync_incremental: bool,
    pub text_document_sync_full: bool,
    pub will_save_provider: bool,
    pub will_save_wait_until_provider: bool,
    pub did_save_provider: bool,
    pub workspace_configuration_provider: bool,
    pub did_change_watched_files_provider: bool,
    pub completion_item_resolve_provider: bool,
    pub code_lens_resolve_provider: bool,
    pub document_link_provider: bool,
    pub document_link_resolve_provider: bool,
    pub color_provider: bool,
    pub declaration_provider: bool,
    pub type_hierarchy_provider: bool,
    pub call_hierarchy_provider: bool,
    pub semantic_tokens_range_provider: bool,
    pub inline_completion_provider: bool,
}

impl ServerCapabilities {
    pub fn from_value(value: &Value) -> Self {
        let Some(obj) = value.as_object() else {
            return ServerCapabilities::default();
        };

        let Some(capabilities) = obj.get("capabilities") else {
            return ServerCapabilities::default();
        };

        let Some(caps_obj) = capabilities.as_object() else {
            return ServerCapabilities::default();
        };

        let mut caps = ServerCapabilities::default();

        caps.hover_provider = is_truthy(caps_obj.get("hoverProvider"));
        caps.signature_help_provider = caps_obj.get("signatureHelpProvider").is_some();
        caps.completion_provider = caps_obj.get("completionProvider").is_some();
        caps.definition_provider = is_truthy(caps_obj.get("definitionProvider"));
        caps.type_definition_provider = is_truthy(caps_obj.get("typeDefinitionProvider"));
        caps.implementation_provider = is_truthy(caps_obj.get("implementationProvider"));
        caps.references_provider = is_truthy(caps_obj.get("referencesProvider"));
        caps.document_symbol_provider = is_truthy(caps_obj.get("documentSymbolProvider"));
        caps.document_formatting_provider = is_truthy(caps_obj.get("documentFormattingProvider"));
        caps.code_action_provider = is_truthy(caps_obj.get("codeActionProvider"));
        caps.rename_provider = is_truthy(caps_obj.get("renameProvider"));
        caps.prepare_rename_provider = is_truthy(caps_obj.get("prepareRenameProvider"));
        caps.inlay_hint_provider = is_truthy(caps_obj.get("inlayHintProvider"));
        caps.code_lens_provider = is_truthy(caps_obj.get("codeLensProvider"));
        caps.semantic_tokens_provider = caps_obj.get("semanticTokensProvider").is_some();
        caps.folding_range_provider = caps_obj.get("foldingRangeProvider").is_some();
        caps.linked_editing_range_provider = caps_obj.get("linkedEditingRangeProvider").is_some();
        caps.selection_range_provider = caps_obj.get("selectionRangeProvider").is_some();
        caps.document_highlight_provider = is_truthy(caps_obj.get("documentHighlightProvider"));
        caps.publish_diagnostics_provider = caps_obj.get("publishDiagnosticsProvider").is_some();
        caps.execute_command_provider = caps_obj.get("executeCommandProvider").is_some();
        caps.did_change_configuration_provider = caps_obj
            .get("workspace")
            .and_then(|w| w.as_object())
            .and_then(|w| w.get("workspaceFolders"))
            .map(|wf| !wf.is_null())
            .unwrap_or(false);
        caps.did_change_workspace_folders_provider = caps_obj
            .get("workspace")
            .and_then(|w| w.as_object())
            .and_then(|w| w.get("workspaceFolders"))
            .map(|wf| !wf.is_null())
            .unwrap_or(false);

        if let Some(ws) = caps_obj.get("workspace") {
            if let Some(ws_obj) = ws.as_object() {
                if let Some(sym) = ws_obj.get("symbol") {
                    caps.workspace_symbol_provider = !sym.is_null();
                }
                if let Some(cfg) = ws_obj.get("fileOperations") {
                    caps.workspace_configuration_provider = !cfg.is_null();
                }
            }
        }
        if let Some(sync) = caps_obj.get("textDocumentSync") {
            match sync {
                Value::Number(n) => {
                    let kind = n.as_u64().unwrap_or(0);
                    caps.text_document_sync_full = kind == 1 || kind == 2;
                    caps.text_document_sync_incremental = kind == 2;
                    caps.will_save_provider = kind > 0;
                    caps.will_save_wait_until_provider = kind > 0;
                    caps.did_save_provider = kind > 0;
                }
                Value::Object(sync_obj) => {
                    let change = sync_obj.get("change").and_then(|v| v.as_u64()).unwrap_or(0);
                    caps.text_document_sync_full = change == 1 || change == 2;
                    caps.text_document_sync_incremental = change == 2;
                    caps.will_save_provider = sync_obj
                        .get("willSave")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false);
                    caps.will_save_wait_until_provider = sync_obj
                        .get("willSaveWaitUntil")
                        .and_then(|v| v.as_bool())
                        .unwrap_or(false);
                    caps.did_save_provider = sync_obj
                        .get("save")
                        .map(|v| !v.is_null() && v != &Value::Bool(false))
                        .unwrap_or(false);
                }
                _ => {}
            }
        }

        caps.did_change_watched_files_provider = caps_obj
            .get("workspace")
            .and_then(|w| w.as_object())
            .and_then(|w| w.get("didChangeWatchedFiles"))
            .map(|v| !v.is_null())
            .unwrap_or(false);

        if !caps.workspace_configuration_provider {
            caps.workspace_configuration_provider = caps_obj
                .get("workspace")
                .and_then(|w| w.as_object())
                .and_then(|w| w.get("workspaceCapabilities"))
                .map(|v| !v.is_null())
                .unwrap_or(false);
        }

        caps.completion_item_resolve_provider = caps_obj
            .get("completionProvider")
            .and_then(|cp| cp.as_object())
            .and_then(|cp| cp.get("resolveProvider"))
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        caps.code_lens_resolve_provider = caps_obj
            .get("codeLensProvider")
            .and_then(|cl| cl.as_object())
            .and_then(|cl| cl.get("resolveProvider"))
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        caps.document_link_provider = caps_obj.get("documentLinkProvider").is_some()
            && !caps_obj["documentLinkProvider"].is_null();
        caps.document_link_resolve_provider = caps_obj
            .get("documentLinkProvider")
            .and_then(|dl| dl.as_object())
            .and_then(|dl| dl.get("resolveProvider"))
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        caps.color_provider = caps_obj
            .get("colorProvider")
            .map(|v| !v.is_null() && v != &Value::Bool(false))
            .unwrap_or(false);

        caps.declaration_provider = is_truthy(caps_obj.get("declarationProvider"));

        caps.type_hierarchy_provider = caps_obj
            .get("typeHierarchyProvider")
            .map(|v| !v.is_null() && v != &Value::Bool(false))
            .unwrap_or(false);

        caps.call_hierarchy_provider = caps_obj
            .get("callHierarchyProvider")
            .map(|v| !v.is_null() && v != &Value::Bool(false))
            .unwrap_or(false);

        caps.semantic_tokens_range_provider = caps_obj
            .get("semanticTokensProvider")
            .and_then(|st| st.as_object())
            .and_then(|st| st.get("range"))
            .map(|v| !v.is_null() && v != &Value::Bool(false))
            .unwrap_or(false);

        caps.inline_completion_provider = caps_obj
            .get("inlineCompletionProvider")
            .map(|v| !v.is_null() && v != &Value::Bool(false))
            .unwrap_or(false);

        caps
    }
}

fn is_truthy(value: Option<&Value>) -> bool {
    match value {
        Some(Value::Bool(b)) => *b,
        Some(Value::Object(_)) => true,
        Some(Value::Null) => false,
        _ => false,
    }
}

pub struct Client {
    child: Child,
    next_id: i64,
    read_buf: Vec<u8>,
}

impl Client {
    pub fn init(server_path: &str, server_args: &[String]) -> std::io::Result<Self> {
        let mut cmd = Command::new(server_path);
        cmd.args(server_args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());

        let child = cmd.spawn()?;

        Ok(Client {
            child,
            next_id: 1,
            read_buf: Vec::new(),
        })
    }

    pub fn send_request(&mut self, method: &str, params: Option<Value>) -> std::io::Result<i64> {
        if let Ok(Some(status)) = self.child.try_wait() {
            let code = status.code().unwrap_or(-1);
            return Err(std::io::Error::new(
                std::io::ErrorKind::BrokenPipe,
                format!("LSP server exited unexpectedly with code {}", code),
            ));
        }

        let id = self.next_id;
        self.next_id += 1;

        let request = JsonRpcRequest {
            jsonrpc: "2.0".to_string(),
            id,
            method: method.to_string(),
            params,
        };

        let content = serde_json::to_string(&request)?;
        self.send_raw(&content)?;
        Ok(id)
    }

    pub fn send_notification(
        &mut self,
        method: &str,
        params: Option<Value>,
    ) -> std::io::Result<()> {
        if let Ok(Some(status)) = self.child.try_wait() {
            let code = status.code().unwrap_or(-1);
            return Err(std::io::Error::new(
                std::io::ErrorKind::BrokenPipe,
                format!("LSP server exited unexpectedly with code {}", code),
            ));
        }

        let notification = JsonRpcNotification {
            jsonrpc: "2.0".to_string(),
            method: method.to_string(),
            params,
        };

        let content = serde_json::to_string(&notification)?;
        self.send_raw(&content)?;
        Ok(())
    }

    fn send_raw(&mut self, content: &str) -> std::io::Result<()> {
        if let Ok(Some(_)) = self.child.try_wait() {
            return Err(std::io::Error::new(
                std::io::ErrorKind::BrokenPipe,
                "LSP server process exited",
            ));
        }

        let stdin = self.child.stdin.as_mut().ok_or(std::io::Error::new(
            std::io::ErrorKind::BrokenPipe,
            "No stdin",
        ))?;

        let header = format!("Content-Length: {}\r\n\r\n", content.len());

        stdin.write_all(header.as_bytes())?;
        stdin.write_all(content.as_bytes())?;
        stdin.flush()?;
        Ok(())
    }

    pub fn read_message(&mut self, timeout: Duration) -> std::io::Result<Option<Value>> {
        let deadline = Instant::now() + timeout;

        while Instant::now() < deadline {
            if let Some(msg) = self.try_parse_message()? {
                return Ok(Some(msg));
            }

            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                return Ok(None);
            }

            let process_exited = match self.child.try_wait() {
                Ok(Some(_)) => true,
                Ok(None) => false,
                Err(_) => false,
            };

            if let Some(stdout) = self.child.stdout.as_mut() {
                let mut tmp = [0u8; 4096];

                match stdout.read(&mut tmp) {
                    Ok(0) => {
                        return Ok(None);
                    }
                    Ok(n) => {
                        self.read_buf.extend_from_slice(&tmp[0..n]);
                        if process_exited {
                            continue;
                        }
                    }
                    Err(ref e) if e.kind() == std::io::ErrorKind::TimedOut => {
                        std::thread::sleep(Duration::from_millis(10));
                    }
                    Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                        std::thread::sleep(Duration::from_millis(10));
                    }
                    Err(_) => {
                        if process_exited {
                            return Ok(None);
                        }
                        return Ok(None);
                    }
                }
            } else {
                return Ok(None);
            }
        }

        Ok(None)
    }

    fn try_parse_message(&mut self) -> std::io::Result<Option<Value>> {
        let data = &self.read_buf[..];
        let sep = match memmem::find(data, b"\r\n\r\n") {
            Some(pos) => pos,
            None => {
                return Ok(None);
            }
        };

        let header = String::from_utf8_lossy(&data[..sep]);
        let cl_start = header.find("Content-Length: ").ok_or_else(|| {
            std::io::Error::new(std::io::ErrorKind::InvalidData, "No Content-Length")
        })?;

        let cl_val_start = cl_start + 16; // "Content-Length: ".len()
        let cl_end = header[cl_val_start..]
            .find('\r')
            .map(|i| cl_val_start + i)
            .unwrap_or(header.len());

        let content_length: usize = header[cl_val_start..cl_end]
            .trim()
            .parse()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;

        let body_start = sep + 4;
        if data.len() < body_start + content_length {
            return Ok(None);
        }

        let body = &data[body_start..body_start + content_length];
        let value: Value = serde_json::from_slice(body)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;

        self.read_buf.drain(..body_start + content_length);

        Ok(Some(value))
    }

    pub fn read_response(&mut self, id: i64, timeout: Duration) -> std::io::Result<Option<Value>> {
        let deadline = Instant::now() + timeout;

        while Instant::now() < deadline {
            let remaining = deadline.saturating_duration_since(Instant::now());
            let msg = match self.read_message(remaining)? {
                Some(m) => m,
                None => {
                    return Ok(None);
                }
            };

            let Some(obj) = msg.as_object() else {
                continue;
            };

            let Some(msg_id) = obj.get("id") else {
                continue;
            };

            let Some(msg_id_int) = msg_id.as_i64() else {
                continue;
            };

            if msg_id_int == id {
                return Ok(Some(msg));
            }
        }

        Ok(None)
    }

    pub fn deinit(&mut self) {
        let _ = self.child.kill();
        let _ = self.child.wait();
    }

    pub fn is_alive(&mut self) -> bool {
        match self.child.try_wait() {
            Ok(None) => true,
            _ => false,
        }
    }
}
